### ISSUE TITLE
CosineEmbeddingCriterion does not minimize Error when using batches

### ISSUE THREAD

**Issue by HarrieO** [torch/nn#287]
The CosineEmbeddingCriterion does not minimize the loss function when using a positive learning rate, surprisingly it does so with a negative learning rate, when using batches.

Simple test case:

      batch_size = 5
      x = torch.randn(batch_size,600)
      y = torch.randn(batch_size,600)
      targets = torch.ones(batch_size)
      data = {x,y}
      for learning_rate = 0.01,-0.01,-0.02 do
         par = nn.ParallelTable()
         par:add(nn.Linear(600,140))
         par:add(nn.Linear(600,140))
         dist = nn.CosineDistance()
         criterion = nn.CosineEmbeddingCriterion(0.5)
         print("Learning Rate", learning_rate)
         for i=1,10 do
            local output = par:forward(data)
            local err = criterion:forward(output, targets)
            print("Cosine Distance",  dist:forward(output)[1])
            print("Error", err)
            local gradOutput = criterion:backward(output, targets)
            par:zeroGradParameters();
            par:backward(data,gradOutput)
            par:updateParameters(learning_rate)
         end
         print(" ")
         print(" ")
      end

Since all targets are one, the Cosine similarity of the outputted vectors should increase and the error decrease. However, the output will be something similar to:

        Learning Rate   0.01    
        Cosine Distance 0.029169121882597   
        Error   0.9708308781174 
        Cosine Distance -0.024496047802374  
        Error   1.0244960478024 
        Cosine Distance -0.078061853162227  
        Error   1.0780618531622 
        Cosine Distance -0.13121970591368   
        Error   1.1312197059137 
        Cosine Distance -0.18366969454385   
        Error   1.1836696945438 
        Cosine Distance -0.2351273055561    
        Error   1.2351273055561 
        Cosine Distance -0.28532943198931   
        Error   1.2853294319893 
        Cosine Distance -0.3340394090272    
        Error   1.3340394090272 
        Cosine Distance -0.38105089136842   
        Error   1.3810508913684 
        Cosine Distance -0.42619047171361   
        Error   1.4261904717136 
            
            
        Learning Rate   -0.01   
        Cosine Distance -0.025475407759606  
        Error   1.0254754077596 
        Cosine Distance 0.024954875095905   
        Error   0.9750451249041 
        Cosine Distance 0.075290123066834   
        Error   0.92470987693317    
        Cosine Distance 0.12527409337952    
        Error   0.87472590662048    
        Cosine Distance 0.17465721134064    
        Error   0.82534278865936    
        Cosine Distance 0.2232015263254 
        Error   0.7767984736746 
        Cosine Distance 0.27068518898849    
        Error   0.72931481101151    
        Cosine Distance 0.31690627635888    
        Error   0.68309372364112    
        Cosine Distance 0.36168583626271    
        Error   0.63831416373729    
        Cosine Distance 0.40487007373161    
        Error   0.59512992626839

The learning rate of -0.01 shows the correct trend, whereas the positive learning rate 0.01 increases the error (and decreases the Cosine distance). Changing the example by making the targets -1 or a mixture of 1 and -1 will not change its behavior.

**Comment by nicholas-leonard** [comment]
Could you send a Pull Request?

**Comment by fmassa** [comment]
If I understand correctly, both `CosineDistance` and `CosineEmbeedingCriterion` were not yet updated to support batches.  Thus, as in your exemple you are using batches, it's probably the reason why you are experiencing this weird behavior. Try replacing by `x` and `y` by `torch.randn (600)` to see if it works.

**Comment by HarrieO** [comment]
@fmassa You're right, when I remove the batch dimension it does behave as expected.

**Comment by fmassa** [comment]
I agree with you that the current implementation is misleading when using batches (no errors or warnings, simply gradients being computed wrongly). I'll try to find some time this week to fix it to accept batch mode.
